### ISSUE TITLE
nutdrv_qx: memory leak

### ISSUE THREAD

**Issue by Yellow-Horror** [networkupstools/nut#2309]
In nutdrv_qx version 2023.07.11 on XigmaNAS (FreeBSD based) there is a memory leak, somehow related to the loss of communication with the UPS.

Symptoms: When communication with the UPS is lost, the memory consumption of the nutdrv_qx process increases by 4 kb every few seconds. Consumed memory will never be released by the process, even if the connection is later restored.

**Comment by jimklimov** [comment]
Thanks for the report! Are you in position to check if this also happens with current NUT codebase (a build from github master) to make sure this was not solved since?

A blind guess would be that in case of connection loss the loop allocates some string to print error messages and never frees it, or something of the sort...

**Comment by Yellow-Horror** [comment]
Unfortunately, I cannot check the relevance of the problem with the latest code, since I am using the “embedded” installation of XigmaNAS, which does not provide the possibility of changing system components. I can promise to check for the problem in the next XigmaNAS release, but the project is developing slowly and does not have a regular release schedule.

**Comment by jimklimov** [comment]
Finally got to glance at the code, and it seems there are two situations that can lead to reconnections, with some code paths (and/or platform depenent builds) maybe mis-balancing `libusb_open()` attempts with `libusb_close()` clean-up. This seems to be a pattern in several USB-capable drivers after a recent-ish refactoring pass.

Maybe the library re-allocates something for the `udev` handle upon each `open()`?.. Didn't quickly find examples or docs stating if it should be `free()`'d by caller, either.

**Comment by jimklimov** [comment]
Ok, tried and failed to reproduce with `usbhid-ups` on Linux, ad least, also after commenting away `ifdef`s that avoided `libusb_close()` on some OSes...

**Comment by jimklimov** [comment]
I don't have a true QX UPS on hand, so hacked `nutdrv_qx.c` to accept whatever device I had to readily plug around. 
* https://github.com/jimklimov/nut/tree/issue-2309-hacks
* `make -s && valgrind --leak-check=full  --show-leak-kinds=all ./drivers/nutdrv_qx -s test -x port=auto -x vendorid=0463 -x productid=ffff -x subdriver=null -DD`


Reconnection attempts after the first disconnection did increase the memory footprint by a few KB (12 to ~100 between runs) according to `top` and `getrusage()`, but the number usually did not move (rarely decreased) after plugging the device back into same or different port.

Tracing with `valgrind` reports a number of allocations remaining not freed, mostly related to dynamic loading and some to libusb. It may well be that the library misbehaves differently on Linux and FreeBSD, but there's little NUT can do about it.

There's also a suspect possible re-allocation of USB matcher structures (exact/regex); can at least tackle those...

````
^C  52.698408   [D1] set_exit_flag: raising exit flag due to signal 2
  52.699204     Signal 2: exiting
  52.700571     [D1] upsdrv_cleanup...
==3825797==
==3825797== HEAP SUMMARY:
==3825797==     in use at exit: 8,272 bytes in 55 blocks
==3825797==   total heap usage: 12,848 allocs, 12,793 frees, 6,327,228 bytes allocated
==3825797==
==3825797== 4 bytes in 1 blocks are still reachable in loss record 1 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4B1BB7A: strdup (strdup.c:42)
==3825797==    by 0x4EC42F6: udev_monitor_filter_add_match_subsystem_devtype (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4A6481B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 11 bytes in 1 blocks are still reachable in loss record 2 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4B1BB7A: strdup (strdup.c:42)
==3825797==    by 0x4EC430F: udev_monitor_filter_add_match_subsystem_devtype (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4A6481B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 15 bytes in 1 blocks are possibly lost in loss record 3 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4B1BB7A: strdup (strdup.c:42)
==3825797==    by 0x13CA4B: xstrdup (common.c:1884)
==3825797==    by 0x119C1B: splitxarg (main.c:1406)
==3825797==    by 0x119C1B: main (main.c:1894)
==3825797==
==3825797== 16 bytes in 1 blocks are still reachable in loss record 4 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4EBF4EB: udev_new (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4A647D6: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 24 bytes in 1 blocks are still reachable in loss record 5 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4EC3C2F: udev_monitor_new_from_netlink (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4A647F5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 24 bytes in 1 blocks are possibly lost in loss record 6 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4A5E48A: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A5E7E2: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59CA6: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 24 bytes in 1 blocks are possibly lost in loss record 7 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4A5E48A: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A5E831: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59CA6: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 33 bytes in 1 blocks are still reachable in loss record 8 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x401C29A: strdup (strdup.c:42)
==3825797==    by 0x40172FB: _dl_load_cache_lookup (dl-cache.c:338)
==3825797==    by 0x4009776: _dl_map_object (dl-load.c:2102)
==3825797==    by 0x400DDC0: openaux (dl-deps.c:64)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x400E138: _dl_map_object_deps (dl-deps.c:248)
==3825797==    by 0x4013DAA: dl_open_worker (dl-open.c:571)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==
==3825797== 33 bytes in 1 blocks are still reachable in loss record 9 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x400BFB7: _dl_new_object (dl-object.c:196)
==3825797==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==3825797==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==3825797==    by 0x400DDC0: openaux (dl-deps.c:64)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x400E138: _dl_map_object_deps (dl-deps.c:248)
==3825797==    by 0x4013DAA: dl_open_worker (dl-open.c:571)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==
==3825797== 38 bytes in 8 blocks are possibly lost in loss record 10 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4B1BB7A: strdup (strdup.c:42)
==3825797==    by 0x4A61714: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==
==3825797== 39 bytes in 3 blocks are definitely lost in loss record 11 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4B1BB7A: strdup (strdup.c:42)
==3825797==    by 0x13CA4B: xstrdup (common.c:1884)
==3825797==    by 0x119C1B: splitxarg (main.c:1406)
==3825797==    by 0x119C1B: main (main.c:1894)
==3825797==
==3825797== 42 bytes in 1 blocks are still reachable in loss record 12 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x401C29A: strdup (strdup.c:42)
==3825797==    by 0x40172FB: _dl_load_cache_lookup (dl-cache.c:338)
==3825797==    by 0x4009776: _dl_map_object (dl-load.c:2102)
==3825797==    by 0x4013D41: dl_open_worker (dl-open.c:513)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==3825797==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==3825797==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==3825797==
==3825797== 42 bytes in 1 blocks are still reachable in loss record 13 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x400BFB7: _dl_new_object (dl-object.c:196)
==3825797==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==3825797==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==3825797==    by 0x4013D41: dl_open_worker (dl-open.c:513)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==3825797==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==3825797==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==3825797==
==3825797== 56 bytes in 1 blocks are still reachable in loss record 15 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x4ECE20E: ??? (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4EC4378: udev_monitor_filter_add_match_subsystem_devtype (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4A6481B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 128 bytes in 8 blocks are possibly lost in loss record 16 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4A619A5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 328 bytes in 1 blocks are still reachable in loss record 17 of 24
==3825797==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==3825797==    by 0x4ED3670: ??? (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4EC3C1F: udev_monitor_new_from_netlink (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==3825797==    by 0x4A647F5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 336 bytes in 1 blocks are possibly lost in loss record 18 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x4012D16: allocate_dtv (dl-tls.c:343)
==3825797==    by 0x4012D16: _dl_allocate_tls (dl-tls.c:589)
==3825797==    by 0x4A78B81: allocate_stack (allocatestack.c:622)
==3825797==    by 0x4A78B81: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==3825797==    by 0x4A648CF: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x12067B: upsdrv_initups (nutdrv_qx.c:3382)
==3825797==    by 0x11A748: main (main.c:2230)
==3825797==
==3825797== 560 bytes in 1 blocks are still reachable in loss record 19 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x4A59AFA: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 744 bytes in 2 blocks are still reachable in loss record 20 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x40116B6: _dl_check_map_versions (dl-version.c:274)
==3825797==    by 0x4013DF5: dl_open_worker (dl-open.c:577)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==3825797==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==3825797==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==3825797==    by 0x4BAF295: nss_load_library (nsswitch.c:359)
==3825797==    by 0x4BAFB28: __nss_lookup_function (nsswitch.c:467)
==3825797==
==3825797== 1,195 bytes in 1 blocks are still reachable in loss record 21 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x400BCDF: _dl_new_object (dl-object.c:89)
==3825797==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==3825797==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==3825797==    by 0x400DDC0: openaux (dl-deps.c:64)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x400E138: _dl_map_object_deps (dl-deps.c:248)
==3825797==    by 0x4013DAA: dl_open_worker (dl-open.c:571)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==
==3825797== 1,204 bytes in 1 blocks are still reachable in loss record 22 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x400BCDF: _dl_new_object (dl-object.c:89)
==3825797==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==3825797==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==3825797==    by 0x4013D41: dl_open_worker (dl-open.c:513)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x40138F9: _dl_open (dl-open.c:837)
==3825797==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==3825797==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==3825797==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==3825797==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==3825797==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==3825797==
==3825797== 1,280 bytes in 8 blocks are possibly lost in loss record 23 of 24
==3825797==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==3825797==    by 0x4A582E5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A63D11: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==    by 0x11A605: main (main.c:2491)
==3825797==
==3825797== 2,048 bytes in 8 blocks are possibly lost in loss record 24 of 24
==3825797==    at 0x48386AF: malloc (vg_replace_malloc.c:306)
==3825797==    by 0x483ADE7: realloc (vg_replace_malloc.c:834)
==3825797==    by 0x4A617ED: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==3825797==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==3825797==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==3825797==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==3825797==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==3825797==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==3825797==
==3825797== LEAK SUMMARY:
==3825797==    definitely lost: 39 bytes in 3 blocks
==3825797==    indirectly lost: 0 bytes in 0 blocks
==3825797==      possibly lost: 3,893 bytes in 36 blocks
==3825797==    still reachable: 4,292 bytes in 15 blocks
==3825797==         suppressed: 48 bytes in 1 blocks
==3825797==
==3825797== For lists of detected and suppressed errors, rerun with: -s
==3825797== ERROR SUMMARY: 9 errors from 9 contexts (suppressed: 0 from 0)
==3825797== could not unlink /tmp/vgdb-pipe-from-vgdb-to-3825797-by-root-on-???
==3825797== could not unlink /tmp/vgdb-pipe-to-vgdb-from-3825797-by-root-on-???
==3825797== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-3825797-by-root-on-???
````

**Comment by jimklimov** [comment]
Tried tracing with just that "hack" commit, and physical reconnections (not simulated by libusb pipe error), and without the potential fix for matcher re-allocation - can not say matchers leak. The set of errors this time is different (maybe Ctrl+C landed in the middle of reading) but all paths lead into libusb:
````
^C  65.443552   [D1] set_exit_flag: raising exit flag due to signal 2
  65.444538     Signal 2: exiting
  65.446274     [D1] upsdrv_cleanup...
==1395203==
==1395203== HEAP SUMMARY:
==1395203==     in use at exit: 8,836 bytes in 61 blocks
==1395203==   total heap usage: 12,699 allocs, 12,638 frees, 6,112,973 bytes allocated
==1395203==
==1395203== 4 bytes in 1 blocks are still reachable in loss record 1 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4B1BB7A: strdup (strdup.c:42)
==1395203==    by 0x4EC42F6: udev_monitor_filter_add_match_subsystem_devtype (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4A6481B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 4 bytes in 1 blocks are possibly lost in loss record 2 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4B1BB7A: strdup (strdup.c:42)
==1395203==    by 0x4A61714: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A6409F: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A645B9: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64716: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A77EA6: start_thread (pthread_create.c:477)
==1395203==    by 0x4B8DA2E: clone (clone.S:95)
==1395203==
==1395203== 11 bytes in 1 blocks are still reachable in loss record 3 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4B1BB7A: strdup (strdup.c:42)
==1395203==    by 0x4EC430F: udev_monitor_filter_add_match_subsystem_devtype (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4A6481B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 15 bytes in 1 blocks are possibly lost in loss record 4 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4B1BB7A: strdup (strdup.c:42)
==1395203==    by 0x13CA0B: xstrdup (common.c:1884)
==1395203==    by 0x119C1B: splitxarg (main.c:1406)
==1395203==    by 0x119C1B: main (main.c:1894)
==1395203==
==1395203== 16 bytes in 1 blocks are still reachable in loss record 5 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4EBF4EB: udev_new (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4A647D6: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 24 bytes in 1 blocks are still reachable in loss record 6 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4EC3C2F: udev_monitor_new_from_netlink (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4A647F5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 24 bytes in 1 blocks are still reachable in loss record 7 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x4A5F62B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5D21B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5E2C6: libusb_handle_events_timeout_completed (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5E37F: libusb_handle_events_completed (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5EE37: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5F109: libusb_control_transfer (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5BB69: libusb_get_string_descriptor_ascii (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x124390: nut_libusb_open (libusb1.c:331)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==
==1395203== 24 bytes in 1 blocks are possibly lost in loss record 8 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4A5E48A: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5E7E2: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59CA6: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 24 bytes in 1 blocks are possibly lost in loss record 9 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4A5E48A: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A5E831: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59CA6: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 24 bytes in 1 blocks are possibly lost in loss record 10 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4A5E48A: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A643B4: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A58DB1: libusb_open (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123C1D: nut_libusb_open (libusb1.c:241)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 33 bytes in 1 blocks are still reachable in loss record 11 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x401C29A: strdup (strdup.c:42)
==1395203==    by 0x40172FB: _dl_load_cache_lookup (dl-cache.c:338)
==1395203==    by 0x4009776: _dl_map_object (dl-load.c:2102)
==1395203==    by 0x400DDC0: openaux (dl-deps.c:64)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x400E138: _dl_map_object_deps (dl-deps.c:248)
==1395203==    by 0x4013DAA: dl_open_worker (dl-open.c:571)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==
==1395203== 33 bytes in 1 blocks are still reachable in loss record 12 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x400BFB7: _dl_new_object (dl-object.c:196)
==1395203==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==1395203==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==1395203==    by 0x400DDC0: openaux (dl-deps.c:64)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x400E138: _dl_map_object_deps (dl-deps.c:248)
==1395203==    by 0x4013DAA: dl_open_worker (dl-open.c:571)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==
==1395203== 38 bytes in 8 blocks are possibly lost in loss record 13 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4B1BB7A: strdup (strdup.c:42)
==1395203==    by 0x4A61714: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==
==1395203== 39 bytes in 3 blocks are definitely lost in loss record 14 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4B1BB7A: strdup (strdup.c:42)
==1395203==    by 0x13CA0B: xstrdup (common.c:1884)
==1395203==    by 0x119C1B: splitxarg (main.c:1406)
==1395203==    by 0x119C1B: main (main.c:1894)
==1395203==
==1395203== 42 bytes in 1 blocks are still reachable in loss record 15 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x401C29A: strdup (strdup.c:42)
==1395203==    by 0x40172FB: _dl_load_cache_lookup (dl-cache.c:338)
==1395203==    by 0x4009776: _dl_map_object (dl-load.c:2102)
==1395203==    by 0x4013D41: dl_open_worker (dl-open.c:513)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==1395203==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==1395203==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==1395203==
==1395203== 42 bytes in 1 blocks are still reachable in loss record 16 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x400BFB7: _dl_new_object (dl-object.c:196)
==1395203==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==1395203==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==1395203==    by 0x4013D41: dl_open_worker (dl-open.c:513)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==1395203==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==1395203==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==1395203==
==1395203== 56 bytes in 1 blocks are still reachable in loss record 18 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x4ECE20E: ??? (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4EC4378: udev_monitor_filter_add_match_subsystem_devtype (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4A6481B: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 96 bytes in 1 blocks are possibly lost in loss record 19 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4A619A5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A6409F: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A645B9: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64716: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A77EA6: start_thread (pthread_create.c:477)
==1395203==    by 0x4B8DA2E: clone (clone.S:95)
==1395203==
==1395203== 128 bytes in 8 blocks are possibly lost in loss record 20 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4A619A5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 160 bytes in 1 blocks are possibly lost in loss record 21 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x4A582E5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D11: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A6409F: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A645B9: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64716: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A77EA6: start_thread (pthread_create.c:477)
==1395203==    by 0x4B8DA2E: clone (clone.S:95)
==1395203==
==1395203== 256 bytes in 1 blocks are possibly lost in loss record 22 of 30
==1395203==    at 0x48386AF: malloc (vg_replace_malloc.c:306)
==1395203==    by 0x483ADE7: realloc (vg_replace_malloc.c:834)
==1395203==    by 0x4A617ED: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A6409F: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A645B9: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64716: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A77EA6: start_thread (pthread_create.c:477)
==1395203==    by 0x4B8DA2E: clone (clone.S:95)
==1395203==
==1395203== 328 bytes in 1 blocks are still reachable in loss record 23 of 30
==1395203==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==1395203==    by 0x4ED3670: ??? (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4EC3C1F: udev_monitor_new_from_netlink (in /usr/lib/x86_64-linux-gnu/libudev.so.1.7.0)
==1395203==    by 0x4A647F5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 336 bytes in 1 blocks are possibly lost in loss record 24 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x4012D16: allocate_dtv (dl-tls.c:343)
==1395203==    by 0x4012D16: _dl_allocate_tls (dl-tls.c:589)
==1395203==    by 0x4A78B81: allocate_stack (allocatestack.c:622)
==1395203==    by 0x4A78B81: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==1395203==    by 0x4A648CF: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A622BC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x12067B: upsdrv_initups (nutdrv_qx.c:3382)
==1395203==    by 0x11A748: main (main.c:2230)
==1395203==
==1395203== 560 bytes in 1 blocks are still reachable in loss record 25 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x4A59AFA: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 744 bytes in 2 blocks are still reachable in loss record 26 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x40116B6: _dl_check_map_versions (dl-version.c:274)
==1395203==    by 0x4013DF5: dl_open_worker (dl-open.c:577)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==1395203==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==1395203==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==1395203==    by 0x4BAF295: nss_load_library (nsswitch.c:359)
==1395203==    by 0x4BAFB28: __nss_lookup_function (nsswitch.c:467)
==1395203==
==1395203== 1,195 bytes in 1 blocks are still reachable in loss record 27 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x400BCDF: _dl_new_object (dl-object.c:89)
==1395203==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==1395203==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==1395203==    by 0x400DDC0: openaux (dl-deps.c:64)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x400E138: _dl_map_object_deps (dl-deps.c:248)
==1395203==    by 0x4013DAA: dl_open_worker (dl-open.c:571)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==
==1395203== 1,204 bytes in 1 blocks are still reachable in loss record 28 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x400BCDF: _dl_new_object (dl-object.c:89)
==1395203==    by 0x4007255: _dl_map_object_from_fd (dl-load.c:997)
==1395203==    by 0x4009274: _dl_map_object (dl-load.c:2236)
==1395203==    by 0x4013D41: dl_open_worker (dl-open.c:513)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x40138F9: _dl_open (dl-open.c:837)
==1395203==    by 0x4BC6F0C: do_dlopen (dl-libc.c:96)
==1395203==    by 0x4BC7A8F: _dl_catch_exception (dl-error-skeleton.c:208)
==1395203==    by 0x4BC7B4E: _dl_catch_error (dl-error-skeleton.c:227)
==1395203==    by 0x4BC6FE6: dlerror_run (dl-libc.c:46)
==1395203==    by 0x4BC7075: __libc_dlopen_mode (dl-libc.c:195)
==1395203==
==1395203== 1,280 bytes in 8 blocks are possibly lost in loss record 29 of 30
==1395203==    at 0x483AB65: calloc (vg_replace_malloc.c:760)
==1395203==    by 0x4A582E5: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D11: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==    by 0x11A605: main (main.c:2491)
==1395203==
==1395203== 2,048 bytes in 8 blocks are possibly lost in loss record 30 of 30
==1395203==    at 0x48386AF: malloc (vg_replace_malloc.c:306)
==1395203==    by 0x483ADE7: realloc (vg_replace_malloc.c:834)
==1395203==    by 0x4A617ED: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A63D36: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A64CAC: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A62196: ??? (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x4A59C93: libusb_init (in /usr/lib/x86_64-linux-gnu/libusb-1.0.so.0.3.0)
==1395203==    by 0x123B2F: nut_libusb_open (libusb1.c:205)
==1395203==    by 0x11C121: qx_command (nutdrv_qx.c:3529)
==1395203==    by 0x120E6D: qx_process (nutdrv_qx.c:4367)
==1395203==    by 0x122041: qx_ups_walk (nutdrv_qx.c:3932)
==1395203==    by 0x122C3C: upsdrv_updateinfo (nutdrv_qx.c:3078)
==1395203==
==1395203== LEAK SUMMARY:
==1395203==    definitely lost: 39 bytes in 3 blocks
==1395203==    indirectly lost: 0 bytes in 0 blocks
==1395203==      possibly lost: 4,433 bytes in 41 blocks
==1395203==    still reachable: 4,316 bytes in 16 blocks
==1395203==         suppressed: 48 bytes in 1 blocks
==1395203==
==1395203== For lists of detected and suppressed errors, rerun with: -s
==1395203== ERROR SUMMARY: 14 errors from 14 contexts (suppressed: 0 from 0)
==1395203== could not unlink /tmp/vgdb-pipe-from-vgdb-to-1395203-by-root-on-???
==1395203== could not unlink /tmp/vgdb-pipe-to-vgdb-from-1395203-by-root-on-???
==1395203== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-1395203-by-root-on-???
````

**Comment by jimklimov** [comment]
I think at the moment this is as far as I can take this quest, without a device and the OS on metal.

If you are able to build NUT on your system (perhaps using a VM with an equivalent FreeBSD release and passing through the USB from UPS), following the `config-prereqs.txt` and building with `ci_build.sh` for quickly wrapped defaults, you can run the driver (as root to avoid USB devfs permissions hassle) under `valgrind` as I outlined above, and reproduce the actual problem with not-hacked driver, maybe those traces would help find a bug in NUT if its code does leak somewhere.

Maybe the actual leak is still in libusb, but some change to call structure from NUT would help (e.g. diligently close/open upon reconnection attempts, take care to `libusb_init()` only once, etc.)?..

**Comment by biergaizi** [comment]
> If you are able to build NUT on your system (perhaps using a VM with an equivalent FreeBSD release and passing through the USB from UPS)

Second this. If one wants to see the problem getting fixed, contributing to the test effort by running the git upstream code is often necessary to accelerate progress (which otherwise may take *years* because of the slow release cycle). If it's inconvenient to test it on your main server, a workaround is to test it on a virtual machine in combination with USB passthrough, or test it on a spare single-board computer like a Raspberry Pi (the last time I've checked, FreeBSD works perfectly fine). These methods allow a quick test without disturbing the server.

**Comment by jimklimov** [comment]
@Yellow-Horror : gentle bump? were you able to test the updated `master` codebase? Note that the PR above should have been released as part of official NUT v2.8.2 too.

**Comment by jimklimov** [comment]
@Yellow-Horror : gentle bump, or I'll close the issue as presumed fixed :D

**Comment by jimklimov** [comment]
Closing for now, feel free to re-open or reference in new issue if the problem is still around after NUT v2.8.2 release.
